### DISCUSSION TITLE
Changing values of an existing NetCDF file (some kind of in-place operation) ?

### DISCUSSION THREAD

**Discussion by jypeter** [xCDAT/xcdat#585] [2. Questions]
### Question criteria

- [X] I added a descriptive title here.
- [X] I searched the [xCDAT GitHub Discussions](https://github.com/xCDAT/xcdat/discussions) to find a similar question and didn't find it.
- [X] I searched the [xCDAT documentation](https://xcdat.readthedocs.io/en/stable/) and/or [Xarray documentation](https://docs.xarray.dev/en/stable/).

### Describe your question

The following question came up in a discussion. It is theoretical, but it was an extremely useful feature of cdms2, when you knew about it. I'm afraid I dont' remember if I have already asked about this

Is there in xarray/xCDAT a feature similar to the _file variables_ (I think) of cdms2 ? I have asked @oliviermarti, but he has told me he only uses xarray/xCDAT to read files, not to create new files or modify existing files

More precisely, I wonder if there is a way to:
- open an existing NetCDF file (or a copy, to be safe) in Read/Write mode
- select a variable
- possibly read a part of the variable using indices/slices
- overwrite parts of the variable in-place using indices, without changing anything else in the file
- close the file

Use cases:
- we have a _Boundary Conditions_ input file with all the correct axes and metadata, and we want to slightly update the BC, without going through the trouble of creating a correct file from scratch. e.g. slightly change a mask, or change some orography data in a region
- we want to correct some values of a variable. e.g we have a file with temperature data and units specified as `K`, but the actual values in the file are in `degC`. So we just want to read the existing `T` data, and overwrite it in-place with `T+273.15`

**Comment by pochedls** [comment]
@jypeter – see [this thread](https://stackoverflow.com/questions/66231575/xarray-appending-or-rewriting-a-existing-nc-file). I think this works a differently in xarray (my understanding/experience is that you essentially need to re-write the whole file/dataset). I think   the netCDF4 package provides this kind of functionality (e.g., [here](https://stackoverflow.com/questions/31865410/python-replacing-values-in-netcdf-file-using-netcdf4)), but you would need to open the netCDF file as a netCDF4 dataset (instead of an xarray dataset).

**Reply by tomvothecoder** [reply]
In addition to the links that @pochedls provided, here's the documentation on [Reading and writing files in Xarray](https://docs.xarray.dev/en/stable/user-guide/io.html) that probably answers your questions.

> Data is always loaded lazily from netCDF files. You can manipulate, slice and subset Dataset and DataArray objects, and no array values are loaded into memory until you try to perform some sort of actual computation. For an example of how these lazy arrays work, see the OPeNDAP section below.

> It is important to note that when you modify values of a Dataset, even one linked to files on disk, only the in-memory copy you are manipulating in xarray is modified: the original file on disk is never touched.

> It is possible to append or overwrite netCDF variables using the `mode='a'` argument. When using this option, all variables in the dataset will be written to the original netCDF file, regardless if they exist in the original dataset.
